pkg/ocm_upgrade_config_manager/ocm_upgrade_config_manager.go

```diff
@@ -0,0 +1,438 @@
+		log.Error(err, "OCM API URL unparsable.")
+		return nil, fmt.Errorf("can't parse OCM API url: %v", err)
+	}
+	url.Path = path.Join(url.Path, CLUSTERS_V1_PATH, cluster.Id, UPGRADEPOLICIES_V1_PATH)
+
+	request := &http.Request{
+		Method: "GET",
+		URL:    url,
+	}
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, fmt.Errorf("can't query upgrade service: %v", err)
+	}
+	operationId := response.Header[OPERATION_ID_HEADER]
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received error code '%v' from OCM upgrade policy service, operation id '%v'", response.StatusCode, operationId)
+	}
+
+	if response.Body != nil {
+		defer response.Body.Close()
+	}
+
+	var upgradeResponse upgradePolicyList
+	decoder := json.NewDecoder(response.Body)
+	decoder.DisallowUnknownFields()
+	err = decoder.Decode(&upgradeResponse)
+	if err != nil {
+		log.Error(err, "unable to decode OCM upgrade policy response, operation id '%v'", operationId)
+		return nil, err
+	}
+
+	return &upgradeResponse, nil
+}
+
+// Applies the supplied Upgrade Policy to the cluster in the form of an UpgradeConfig
+// Returns an indication of if the policy being applied differs to the existing UpgradeConfig,
+// and indication of error if one occurs.
+func applyUpgradePolicies(upgradePolicies *upgradePolicyList, cluster *clusterInfo, c client.Client) (bool, error) {
+	// Fetch the current UpgradeConfig instance, if it exists
+	upgradeConfigs := &upgradev1alpha1.UpgradeConfigList{}
+	err := c.List(context.TODO(), upgradeConfigs, &client.ListOptions{})
+	if err != nil {
+		return false, fmt.Errorf("unable to retrieve upgradeconfigs")
+	}
+
+	// If there are no policies available, or if there is no next upgrade scheduled, do not retain any UpgradeConfigs
+	if len(upgradePolicies.Items) == 0 || upgradePolicies.Items[0].NextRun == "" {
+		if len(upgradeConfigs.Items) > 0 {
+			for _, upgradeConfig := range upgradeConfigs.Items {
+				log.Info(fmt.Sprintf("Removing expired UpgradeConfig %s", upgradeConfig.Name))
+				err = c.Delete(context.TODO(), &upgradeConfig, &client.DeleteOptions{})
+				if err != nil {
+					return false, fmt.Errorf("unable to delete existing upgrade config: %v", err)
+				}
+			}
+		}
+		return true, nil
+	}
+
+	// Only handling singular upgrade policies for now - but log if that arises
+	if len(upgradePolicies.Items) > 1 {
+		log.Info("More than one Upgrade Policy received, only considering the first.")
+	}
+	upgradePolicy := upgradePolicies.Items[0]
+
+	// Check if we have an existing UpgradeConfig to compare against, for the refresh
+	originalUpgradeConfig := upgradev1alpha1.UpgradeConfig{}
+	if len(upgradeConfigs.Items) > 0 {
+		originalUpgradeConfig = upgradeConfigs.Items[0]
+	}
+
+	// Set up an UpgradeConfig that reflects the policy
+	replacementUpgradeConfig := upgradev1alpha1.UpgradeConfig{}
+	// If there was an existing UpgradeConfig, make a clone of its contents
+	originalUpgradeConfig.DeepCopyInto(&replacementUpgradeConfig)
```

Comment:
It is better to move this into the if condition above?

Reply:
Agree, done in e77eca5